camel-core/src/main/java/org/apache/camel/processor/RestBindingAdvice.java

```diff
@@ -196,7 +194,11 @@ private void unmarshal(Exchange exchange, Map<String, Object> state) throws Exce
                 // so force reading the body as a String which we can work with
                 body = MessageHelper.extractBodyAsString(exchange.getIn());
                 if (body != null) {
-                    exchange.getIn().setBody(body);
+                    if (exchange.getIn() instanceof DataTypeAware) {
+                        ((DataTypeAware)exchange.getIn()).setBody(body, new DataType(isJson ? "json" : "xml"));
```

Comment:
isn't risky to suppose that if not json, it is forcibly xml? In this case why do we need the isXml variable?